_posts/2022-05-14-appsync-graphql-subscriptions.md

```diff
@@ -0,0 +1,308 @@
+  updateOrderStatus(orderId: ID!, orderStatus: OrderStatus!): Order!
+}
+
+input orderInput {
+  orderItems: [OrderItems!]!
+  status: OrderStatus!
+  createdAt: String!
+}
+
+type Order {
+  id: String!
+  orderItems: [OrderItems!]!
+  status: OrderStatus!
+  createdAt: String!
+}
+
+type OrderItems {
+  item: String!
+  quantity: Int!
+}
+
+enum OrderStatus {
+  IN_PROGRESS
+  CONFIRMED
+  READY
+  CANCELLED
+}
+```
+
+corresponding GraphGL mutation on Frontend:
+
+```
+export const UPDATE_ORDER_STATUS = gql`
+  mutation UpdateOrderStatus(orderId: ID!, orderStatus: OrderStatus!) {
+    updateOrderStatus(orderId: $orderId, orderStatus: $orderStatus) {
+      id
+      orderItems {
+        item
+        quantity
+      }
+      OrderStatus
+      createdAt
+    }
+  }
+`;
+
+```
+
+As the schema shows, we have a mutation which updates the status of an order in a data source like our database of choice. We want our UI to get the live status of the order, each time an event updates the status.
+
+To enable subscription on any type and to make these fields real time, you can simply create a subscription type in your GraphQL schema and just mention the `@aws_subscribe` directive for subscription you want to receive the changes.That will allow the client to receive a real-time update on order.
+
+#### Step1:
+
+Update your Schema.gql with:
+
+```
+type Subscription {
+    getOrderStatus($orderId: ID!): Order @aws_subscribe(mutations: ["updateOrderStatus"])
+```
+
+#### Step2:
+
+Define the subscription in your Frontend:
+
+```
+export const GET_ORDER_STATUS = gql`
+  subscription GetOrderStatus($orderId: ID) {
+    getOrderStatus(orderId: $orderId) {
+      id
+      orderItems {
+        item
+        quantity
+      }
+      OrderStatus
+      createdAt
+    }
+  }
+`;
+```
+
+#### Step3:
+
+In your UI component,use `useSubscription` Hook to execute a subscription from React. The following example component uses the subscription we defined earlier to render the most recent order status. Whenever the GraphQL server pushes a new status to the client, the component re-renders with the updated order.
+
+```
+---- LatestOrderStatus.tsx file
+
+import { ApolloError } from 'apollo-client';
+import React from 'react';
+import {  useSubscription } from '@apollo/react-hooks';
+
+const LatestOrderStatus = ({ orderId: string }) =>  {
+  const { data, loading } = useSubscription(
+    GET_ORDER_STATUS,
+    { variables: { orderId },
+    onSubscriptionData: order => {
+      // Add your code here for whatever is going to happen when subscriber gets the new order data.
+    }
+     }
+  );
+  return <h4> You order status is: {!loading && data.getOrderStatus.status}</h4>;
+}
+
+export default LatestOrderStatus;
+
+```
+
+Yes, as simple as this we could implement a real time communication between our client and the server. However, there are few lessons I have learned when I tried to implement a subscription for one of our recent projects.
```

Comment:
Yes, as simple as this we could implement a real time communication
could be
It is as simple as this to implement real time communication